CORE-5088 Consensual ledger persistence

mbrkic-r3 opened the pull request:
This PR continues remaining work of https://github.com/corda/corda-runtime-os/pull/1883 CORE-5088: (Start of) consensual ledger persistence

### TODO

- [x] Moved modules "entity-processor-service", "entity-processor-service-impl", "ledger-persistence-service" and "ledger-persistence-service-impl" from components:virtual-node" to "components:persistence"
- [x] Persist transaction signatures
- [x] Refactor to work with ConsensualSignedTransactionImpl instead of WireTransaction
- [x] Persist serialized DigitalSignatureAndMetadata in DB
- [x] Persist public key hash to "consensual_transaction_signature" DB table
- [x] Link transaction only to existing CPKs
- [x] Refactor message processor to return missing CPKs when persisting transaction.
- [x] Return missing CPKs when persisting transaction in ledger-persistence-client
- [ ] Remove dependencies on unnecessary flow components (depends on https://github.com/corda/corda-runtime-os/pull/2321)

mbrkic-r3 committed:
CORE-5088 Consensual ledger persistence service in flow component

mbrkic-r3 committed:
CORE-5088 Consensual ledger persistence service test

mbrkic-r3 committed:
CORE-5088 Removed @Suspendable annotations in LedgerPersistenceServiceImpl

mbrkic-r3 committed:
CORE-5088 Bumped cordaApiVersion

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into mbrkic-r3/core-5088/consensual-ledger-persistence-flow

# Conflicts:
#	gradle.properties

mbrkic-r3 committed:
CORE-5088 Added runtime dependency to libs:crypto:merkle-impl

mbrkic-r3 committed:
CORE-5088 Added service value for @References that missed them

mbrkic-r3 committed:
CORE-5088 Moved modules "entity-processor-service", "entity-processor-service-impl", "ledger-persistence-service" and "ledger-persistence-service-impl" from components:virtual-node" to "components:persistence"

mbrkic-r3 commented:
@driessamyn commented in https://github.com/corda/corda-runtime-os/pull/1883:
> Not been able to review this in detail (will try to later). I left a couple of comments, but the main two things I think I would consider changing:
> 
> * Move packages into a `persistence` namespace, with only the Sandbox things left in vnode. We should be able to break the bit that creates and sets up the Sandboxes from the processors.
> * Too  much native SQL. I can't see a good reason to use native SQL here. I know we talked about this in context of the UTXO ledger where we may want to use RDBM specitic SQL, but what is in here seems very generic and creating this as native SQL only results in a potential tax against future compatibility when introducing other RDBMs.

mbrkic-r3 commented:
@blsemo commented in https://github.com/corda/corda-runtime-os/pull/1883:


> > ```
> > * Too  much native SQL. I can't see a good reason to use native SQL here. I know we talked about this in context of the UTXO ledger where we may want to use RDBM specitic SQL, but what is in here seems very generic and creating this as native SQL only results in a potential tax against future compatibility when introducing other RDBMs.
> > ```
> 
> But that is kind of the point of the processor - I know the native SQL isn't strictly necessary here, but this is supposed to be a simplified blueprint for the UTXO tx processor. Also, the database structure is sufficiently complex that the JPA annotations wouldn't be straight forward, I believe.

mbrkic-r3 committed:
CORE-5088 Added method for persisting signatures

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into mbrkic-r3/core-5088/consensual-ledger-persistence-flow

driessamyn commented:
> @blsemo commented in #1883:
> 
> > > ```
> > > * Too  much native SQL. I can't see a good reason to use native SQL here. I know we talked about this in context of the UTXO ledger where we may want to use RDBM specitic SQL, but what is in here seems very generic and creating this as native SQL only results in a potential tax against future compatibility when introducing other RDBMs.
> > > ```
> > 
> > 
> > But that is kind of the point of the processor - I know the native SQL isn't strictly necessary here, but this is supposed to be a simplified blueprint for the UTXO tx processor. Also, the database structure is sufficiently complex that the JPA annotations wouldn't be straight forward, I believe.

I understand that this is a blueprint for the UTXO processor and that we may want to prove we can use native SQL, but I still think we're over-using it (or at least inconsistently with the rest of the codebase). My view would be that we only use NativeSQL where JPQL or JPA queries aren't an option or too much a compromise. Whenever we have a native query, there is a chance there will need to be variations for different DB vendors and from what I can see these are all pretty straightforward joins, so I don't think we're getting a big benefit.

I'm not completely opposed to it, but I am concerned about yet more inconsistency in the codebase and in my opinion we should have clear guidelines of when we use JPA and when we use native queries as I can see some people would look at this and want to replace some of the existing, or new, structures with native SQL because on the surface it looks simpler and quicker.

mbrkic-r3 committed:
CORE-5088 Refactored to address review comments

mbrkic-r3 committed:
CORE-5088 Refactored to work with ConsensualSignedTransactionImpl instead of WireTransaction

mbrkic-r3 committed:
CORE-5088 Fixed failing integration test

vlajos committed:
CORE-5091 Consensual Signed Tx signature tweaks

omfR3 committed:
CORE-5940: (Consensual ledger) Add mocks for recently-added services used by ConsensualLedgerServiceImpl and ConsensualLedgerTransactionImpl. Tests still failing.

omfR3 committed:
CORE-5940: (Consensual ledger) Add helper for mocks relating to transactions and tx builder + temporarily roll back some of the tx builder signing changes

mbrkic-r3 committed:
CORE-5088 Persisting serialized DigitalSignatureAndMetadata in DB

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into mbrkic-r3/core-5088/consensual-ledger-persistence-flow

mbrkic-r3 committed:
CORE-5088 Fixed compile errors after using api with renamed messaging schemas

mbrkic-r3 committed:
CORE-5088 Reverted changes in flow-service build.gradle

mbrkic-r3 committed:
CORE-5088 Fixed compile errors in tests

omfR3 committed:
CORE-5940: (ledger-common) Add some CPK information (name and checksum) to transaction metadata

omfR3 committed:
CORE-5940: (ledger-common/consensual) Add missing mocks to the other consensual ledger TX tests + handle null CPK list in metadata

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

# Conflicts:
#	components/persistence/ledger-persistence-service-impl/build.gradle
#	components/persistence/ledger-persistence-service-impl/src/integrationTest/kotlin/net/corda/ledger/persistence/impl/tests/ConsensualLedgerMessageProcessorTests.kt
#	components/persistence/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerMessageProcessor.kt
#	components/persistence/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerRepository.kt

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/mbrkic-r3/core-5088/consensual-ledger-persistence-flow' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

mbrkic-r3 committed:
CORE-5088 Persisting public key hash to "consensual_transaction_signature" DB table

mbrkic-r3 committed:
CORE-5088 Refactoring

mbrkic-r3 committed:
CORE-5088 Bumped API version

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

# Conflicts:
#	gradle.properties

mbrkic-r3 committed:
CORE-5088 Bumped API version

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into mbrkic-r3/core-5088/consensual-ledger-persistence-flow

# Conflicts:
#	gradle.properties

omfR3 committed:
CORE-5940: (Consensual ledger) Fix integration test OSGi error

omfR3 committed:
CORE-5940: (Consensual ledger) Add more CPK information to transaction metadata

mbrkic-r3 committed:
CORE-5088 Refactored to work with ConsensualSignedTransactionImpl, renamed module and classes

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/mbrkic-r3/core-5088/consensual-ledger-persistence-flow' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

omfR3 committed:
CORE-5940: Fix detekt errors

lankydan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/factory/impl/FlowRecordFactoryImpl.kt:
```diff
@@ -36,7 +36,7 @@ class FlowRecordFactoryImpl : FlowRecordFactory {
     override fun createEntityRequestRecord(requestId: String, payload: EntityRequest): Record<String, EntityRequest> {
         return Record(
-            topic = ENTITY_PROCESSOR,
+            topic = PERSISTENCE_ENTITY_PROCESSOR_TOPIC,
```
Can you do me a favour and delete `createEntityRequestRecord` and `createFlowOpsRequestRecord`. They are no longer needed due to the external event work I completed, however I did not seem to remove these methods.

mbrkic-r3 replied:
deleted

lankydan commented on components/persistence/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ComponentGroupListsTuplesMapper.kt:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Used by [ConsensualLedgerRepository.findTransaction] to map DB rows to transaction's components group lists
+ */
+class ComponentGroupListsTuplesMapper: TuplesMapper<List<ByteArray>> {
```
Does this need to be a class if it doesn't have any dependencies?

lankydan commented on components/persistence/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerMessageProcessor.kt:
```diff
@@ -115,5 +119,5 @@ class ConsensualLedgerMessageProcessor(
 
     private fun SerializationService.deserialize(persistTransaction: PersistTransaction) =
-        deserialize(persistTransaction.transaction.array(), WireTransaction::class.java)
+        deserialize<ConsensualSignedTransactionImpl>(persistTransaction.transaction.array())
```
Can we reference `ConsensualSignedTransaction` instead of its implementation directly.

mbrkic-r3 replied:
Repository needs `ConsensualSignedTransactionImpl` to be able to persist all data so we would need to cast it to impl again

lankydan commented on components/persistence/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerRepository.kt:
```diff
@@ -53,108 +44,138 @@ class ConsensualLedgerRepository(
+                """,
+            Tuple::class.java)
             .setParameter("id", id)
             .resultList
```
Can we do `resultList as List<Tuple>` instead of mapping each one to a `Tuple`?

mbrkic-r3 replied:
refactored

lankydan commented on components/persistence/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerRepository.kt:
```diff
@@ -53,108 +44,138 @@ class ConsensualLedgerRepository(
         val privacySalt = PrivacySaltImpl(firstRowColumns[1] as ByteArray)
-        val componentGroupLists = queryRowsToComponentGroupLists(rows)
-        return WireTransaction(merkleTreeProvider, digestService, jsonMarshallingService, privacySalt, componentGroupLists)
+        val componentGroupLists = rows.mapTuples(componentGroupListsTuplesMapper)
```
I think `mapTuples(componentGroupListsTuplesMapper)` and its corresponding interface seems a bit weird. Would an extension function of `toComponentGroupLists` or `mapComponentGroupLists` be simpler or is there something I am missing?

lankydan commented on components/persistence/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerRepository.kt:
```diff
@@ -53,108 +44,138 @@ class ConsensualLedgerRepository(
+            Tuple::class.java)
+            .setParameter("transactionId", transactionId)
+            .resultList
+            .map { it as Tuple }
```
Fair I see why you did this in the other code now, guess it keeps it consistent.

Might be worth creating an extension function `resultListAsTuples` or something which can combine this if its common.

mbrkic-r3 replied:
Yes, to keep it consistent as using native queries without entity classes will require working with `Tuples` for retrievals (and we also want logic that maps tuples to domain classes visible so it could be unit tested, hence `TuplesMapper`)
Created `resultListAsTuples`

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into mbrkic-r3/core-5088/consensual-ledger-persistence-flow

mbrkic-r3 committed:
CORE-5088 Refactoring

mbrkic-r3 committed:
CORE-5088 Refactored

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/mbrkic-r3/core-5088/consensual-ledger-persistence-flow' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

omfR3 committed:
CORE-5940: (ledger-common / consensual) Provide some structure to CPI/CPK metadata in transactions rather than mushing it all into a string.

omfR3 committed:
CORE-5940: (ledger-common + consensual) Rename transaction.CpkIdentifier -> transaction.CpkMetadata and use them to encapsulate tx CPK metadata rather than fumbling with ordered maps. Something is still broken.

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

# Conflicts:
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/factory/impl/FlowRecordFactoryImpl.kt
#	gradle.properties
#	libs/ledger/ledger-persistence-client/src/main/kotlin/net/corda/ledger/consensual/persistence/external/events/AbstractLedgerExternalEventFactory.kt
#	libs/ledger/ledger-persistence-client/src/main/kotlin/net/corda/ledger/consensual/persistence/internal/LedgerPersistenceServiceImpl.kt

vlajos committed:
CORE-5091 Instead of using cpi-info-read-service and virtual-node-info-read-service we extract the CPK info directly from sandboxGroupContext.sandboxGroup.metadata

(cherry picked from commit 5929679c5cc81a028b2e6a5806bf1a26ef096b77)

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/CORE-5940/encode-cpk-metadata-in-consensual-transactions' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

# Conflicts:
#	components/virtual-node/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerRepository.kt

omfR3 committed:
CORE-5940: (ledger-common) Reconstitute CpiMetadata after JSON mapper turns it into a generic Map<String, Any?>

omfR3 committed:
CORE-5940: (ledger-common) Refactor CpiSummary + TransactionMetadata

* Hoist list of CPK metadata (CpkSummary) out of CpiSummary
* Add separate CPK_METADATA_KEY + accessor to TransactionMetaData

omfR3 committed:
CORE-5940: (ledger-consensual) Use accessor to get CPK metadata so we can persist it later

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-5940/encode-cpk-metadata-in-consensual-transactions

yift-r3 committed:
CORE-7041: Remove warning from startup

omfR3 committed:
Merge branch 'yift/core-7041/remove-warning' of github.com:corda/corda-runtime-os into CORE-5940/encode-cpk-metadata-in-consensual-transactions

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/CORE-5940/encode-cpk-metadata-in-consensual-transactions' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

# Conflicts:
#	components/virtual-node/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerRepository.kt

omfR3 committed:
CORE-5094: (ledger-consensual-impl) Use List rather than abstract Collection type to get around Kryoserializer error

omfR3 committed:
CORE-5940: (ledger-consensual) Convert Collection<T> -> List<T> more neatly

omfR3 committed:
Merge branch 'release/os/5.0' of github.com:corda/corda-runtime-os into CORE-5940/encode-cpk-metadata-in-consensual-transactions

omfR3 committed:
CORE-5940: (ledger-common) Better CPI/CPK info validation in TX metadata

When converting TX metadata, don't accept e.g. a null version and turn it into the String "null". Check all fields and throw sensible runtime error if any are bad (missing or wrong type).

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/CORE-5940/encode-cpk-metadata-in-consensual-transactions' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

mbrkic-r3 committed:
CORE-5088 Fixed tests after merge. Linking transaction only to existing CPKs. Refactored message processor to return missing CPKs when persisting transaction.

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into mbrkic-r3/core-5088/consensual-ledger-persistence-2

# Conflicts:
#	components/virtual-node/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerRepository.kt
#	gradle.properties
#	libs/ledger/ledger-common-impl/src/main/kotlin/net/corda/ledger/common/impl/transaction/CpkSummary.kt
#	libs/ledger/ledger-common-impl/src/main/kotlin/net/corda/ledger/common/impl/transaction/TransactionMetaData.kt
#	libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualTransactionBuilderImpl.kt
#	libs/ledger/ledger-consensual-impl/src/test/kotlin/net/corda/ledger/consensual/impl/ConsensualTransactionMocks.kt
#	libs/ledger/ledger-consensual-impl/src/test/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualTransactionBuilderImplTest.kt

mbrkic-r3 committed:
CORE-5088 Fixed compile errors after merge

mbrkic-r3 committed:
CORE-5088 Refactoring

mbrkic-r3 committed:
CORE-5088 Returning missing CPKs when persisting transaction in ledger-persistence-client

lankydan commented:
there are too many random commits in this PR to review it :/

mbrkic-r3 commented:
Replaced with https://github.com/corda/corda-runtime-os/pull/2383
